// eslint-plugin-regexp/dist/utils/type-tracker/type-data/bigint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buildBigIntConstructor = exports.BIGINT = exports.TypeBigInt = void 0;
const common_1 = require("./common");
const function_1 = require("./function");
const object_1 = require("./object");
class TypeBigInt {
    constructor() {
        this.type = "BigInt";
    }
    has(type) {
        return type === "BigInt";
    }
    paramType() {
        return null;
    }
    propertyType(name) {
        return getPrototypes()[name] || null;
    }
    iterateType() {
        return null;
    }
    returnType() {
        return null;
    }
    typeNames() {
        return ["BigInt"];
    }
    equals(o) {
        return o.type === "BigInt";
    }
}
exports.TypeBigInt = TypeBigInt;
exports.BIGINT = new TypeBigInt();
function buildBigIntConstructor() {
    const BIGINT_TYPES = common_1.createObject({
        asIntN: function_1.RETURN_BIGINT,
        asUintN: function_1.RETURN_BIGINT,
        prototype: null,
    });
    return new function_1.TypeGlobalFunction(() => exports.BIGINT, BIGINT_TYPES);
}
exports.buildBigIntConstructor = buildBigIntConstructor;
const getPrototypes = common_1.cache(() => common_1.createObject(Object.assign(Object.assign({}, object_1.getObjectPrototypes()), { toString: function_1.RETURN_STRING, toLocaleString: function_1.RETURN_STRING, valueOf: function_1.RETURN_BIGINT })));
