// eslint-plugin-regexp/dist/utils/type-tracker/type-data/global.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GLOBAL = exports.TypeGlobal = void 0;
const array_1 = require("./array");
const bigint_1 = require("./bigint");
const boolean_1 = require("./boolean");
const common_1 = require("./common");
const function_1 = require("./function");
const map_1 = require("./map");
const number_1 = require("./number");
const object_1 = require("./object");
const regexp_1 = require("./regexp");
const set_1 = require("./set");
const string_1 = require("./string");
class TypeGlobal {
    constructor() {
        this.type = "Global";
    }
    has(_type) {
        return false;
    }
    paramType() {
        return null;
    }
    propertyType(name) {
        return getProperties()[name] || null;
    }
    iterateType() {
        return null;
    }
    returnType() {
        return null;
    }
    typeNames() {
        return ["Global"];
    }
    equals(o) {
        return o.type === "Global";
    }
}
exports.TypeGlobal = TypeGlobal;
exports.GLOBAL = new TypeGlobal();
const getProperties = common_1.cache(() => common_1.createObject({
    String: string_1.buildStringConstructor(),
    Number: number_1.buildNumberConstructor(),
    Boolean: boolean_1.buildBooleanConstructor(),
    RegExp: regexp_1.buildRegExpConstructor(),
    BigInt: bigint_1.buildBigIntConstructor(),
    Array: array_1.buildArrayConstructor(),
    Object: object_1.buildObjectConstructor(),
    Function: function_1.buildFunctionConstructor(),
    Map: map_1.buildMapConstructor(),
    Set: set_1.buildSetConstructor(),
    isFinite: function_1.RETURN_BOOLEAN,
    isNaN: function_1.RETURN_BOOLEAN,
    parseFloat: function_1.RETURN_NUMBER,
    parseInt: function_1.RETURN_NUMBER,
    decodeURI: function_1.RETURN_STRING,
    decodeURIComponent: function_1.RETURN_STRING,
    encodeURI: function_1.RETURN_STRING,
    encodeURIComponent: function_1.RETURN_STRING,
    escape: function_1.RETURN_STRING,
    unescape: function_1.RETURN_STRING,
    globalThis: exports.GLOBAL,
    window: exports.GLOBAL,
    self: exports.GLOBAL,
    global: exports.GLOBAL,
    undefined: "undefined",
    Infinity: number_1.NUMBER,
    NaN: number_1.NUMBER,
}));
