// eslint-plugin-regexp/dist/utils/type-tracker/type-data/iterable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UNKNOWN_ITERABLE = exports.TypeIterable = void 0;
const common_1 = require("./common");
const object_1 = require("./object");
const getPrototypes = common_1.cache(() => {
    return common_1.createObject(Object.assign({}, object_1.getObjectPrototypes()));
});
class TypeIterable {
    constructor(param0) {
        this.type = "Iterable";
        this.param0 = param0;
    }
    has(_type) {
        return false;
    }
    paramType(index) {
        if (index === 0) {
            return this.param0();
        }
        return null;
    }
    propertyType(name) {
        return getPrototypes()[name] || null;
    }
    iterateType() {
        return this.paramType(0);
    }
    returnType() {
        return null;
    }
    typeNames() {
        const param0 = common_1.getTypeName(this.iterateType());
        return [`Iterable${param0 != null ? `<${param0}>` : ""}`];
    }
    equals(o) {
        if (o.type !== "Iterable") {
            return false;
        }
        return common_1.isEquals(this.iterateType(), o.iterateType());
    }
}
exports.TypeIterable = TypeIterable;
exports.UNKNOWN_ITERABLE = new TypeIterable(() => null);
