// prettier-plugin-jsdoc/lib/roles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TAGS_VERTICALLY_ALIGN_ABLE = exports.TAGS_TYPELESS = exports.TAGS_TYPE_NEEDED = exports.TAGS_SYNONYMS = exports.TAGS_ORDER = exports.TAGS_GROUP_CONDITION = exports.TAGS_GROUP_HEAD = exports.TAGS_NAMELESS = exports.TAGS_DESCRIPTION_NEEDED = exports.TAGS_PEV_FORMATE_DESCRIPTION = void 0;
const tags_1 = require("./tags");
const TAGS_SYNONYMS = {
    // One TAG TYPE can have different titles called SYNONYMS.  We want
    // to avoid different titles in the same tag so here is map with
    // synonyms as keys and tag type as value that we want to have in
    // final jsDoc.
    arg: tags_1.PARAM,
    argument: tags_1.PARAM,
    const: tags_1.CONSTANT,
    constructor: tags_1.CLASS,
    desc: tags_1.DESCRIPTION,
    emits: tags_1.FIRES,
    examples: tags_1.EXAMPLE,
    exception: tags_1.THROWS,
    fileoverview: tags_1.FILE,
    func: tags_1.FUNCTION,
    host: tags_1.EXTERNAL,
    method: tags_1.FUNCTION,
    overview: tags_1.FILE,
    params: tags_1.PARAM,
    prop: tags_1.PROPERTY,
    return: tags_1.RETURNS,
    var: tags_1.MEMBER,
    virtual: tags_1.ABSTRACT,
    yield: tags_1.YIELDS,
    hidden: tags_1.IGNORE,
};
exports.TAGS_SYNONYMS = TAGS_SYNONYMS;
const TAGS_NAMELESS = [
    tags_1.BORROWS,
    tags_1.CATEGORY,
    tags_1.DEPRECATED,
    tags_1.DESCRIPTION,
    tags_1.EXAMPLE,
    tags_1.EXTENDS,
    tags_1.LICENSE,
    tags_1.MODULE,
    tags_1.NAMESPACE,
    tags_1.OVERRIDE,
    tags_1.PRIVATE_REMARKS,
    tags_1.REMARKS,
    tags_1.RETURNS,
    tags_1.SINCE,
    tags_1.THROWS,
    tags_1.TODO,
    tags_1.YIELDS,
];
exports.TAGS_NAMELESS = TAGS_NAMELESS;
const TAGS_TYPELESS = [
    tags_1.BORROWS,
    tags_1.BORROWS,
    tags_1.DEPRECATED,
    tags_1.DESCRIPTION,
    tags_1.EXAMPLE,
    tags_1.IGNORE,
    tags_1.LICENSE,
    tags_1.MODULE,
    tags_1.NAMESPACE,
    tags_1.OVERRIDE,
    tags_1.PRIVATE_REMARKS,
    tags_1.REMARKS,
    tags_1.SINCE,
    tags_1.TODO,
];
exports.TAGS_TYPELESS = TAGS_TYPELESS;
const TAGS_PEV_FORMATE_DESCRIPTION = [
    /** @todo should be formate like jsdoc standard saw https://jsdoc.app/tags-borrows.html  */
    tags_1.BORROWS,
    tags_1.DEFAULT,
    tags_1.DEFAULT_Value,
    tags_1.MEMBEROF,
    tags_1.MODULE,
    tags_1.SEE,
];
exports.TAGS_PEV_FORMATE_DESCRIPTION = TAGS_PEV_FORMATE_DESCRIPTION;
const TAGS_DESCRIPTION_NEEDED = [
    tags_1.BORROWS,
    tags_1.CATEGORY,
    tags_1.DESCRIPTION,
    tags_1.EXAMPLE,
    tags_1.PRIVATE_REMARKS,
    tags_1.REMARKS,
    tags_1.SINCE,
    tags_1.TODO,
];
exports.TAGS_DESCRIPTION_NEEDED = TAGS_DESCRIPTION_NEEDED;
const TAGS_TYPE_NEEDED = [
    tags_1.EXTENDS,
    tags_1.PARAM,
    tags_1.PROPERTY,
    tags_1.RETURNS,
    tags_1.THROWS,
    tags_1.TYPE,
    tags_1.TYPEDEF,
    tags_1.YIELDS,
];
exports.TAGS_TYPE_NEEDED = TAGS_TYPE_NEEDED;
const TAGS_VERTICALLY_ALIGN_ABLE = [
    tags_1.EXTENDS,
    tags_1.PARAM,
    tags_1.PROPERTY,
    tags_1.RETURNS,
    tags_1.THROWS,
    tags_1.TYPE,
    tags_1.TYPEDEF,
    tags_1.YIELDS,
];
exports.TAGS_VERTICALLY_ALIGN_ABLE = TAGS_VERTICALLY_ALIGN_ABLE;
const TAGS_GROUP_HEAD = [tags_1.CALLBACK, tags_1.TYPEDEF];
exports.TAGS_GROUP_HEAD = TAGS_GROUP_HEAD;
const TAGS_GROUP_CONDITION = [
    ...TAGS_GROUP_HEAD,
    tags_1.TYPE,
    tags_1.PROPERTY,
    tags_1.PARAM,
    tags_1.RETURNS,
    tags_1.YIELDS,
    tags_1.THROWS,
];
exports.TAGS_GROUP_CONDITION = TAGS_GROUP_CONDITION;
const TAGS_ORDER = [
    tags_1.REMARKS,
    tags_1.PRIVATE_REMARKS,
    tags_1.PROVIDES_MODULE,
    tags_1.MODULE,
    tags_1.LICENSE,
    tags_1.FLOW,
    tags_1.ASYNC,
    tags_1.PRIVATE,
    tags_1.IGNORE,
    tags_1.MEMBEROF,
    tags_1.VERSION,
    tags_1.AUTHOR,
    tags_1.DEPRECATED,
    tags_1.SINCE,
    tags_1.CATEGORY,
    tags_1.DESCRIPTION,
    tags_1.EXAMPLE,
    tags_1.ABSTRACT,
    tags_1.AUGMENTS,
    tags_1.CONSTANT,
    tags_1.DEFAULT,
    tags_1.DEFAULT_Value,
    tags_1.EXTERNAL,
    tags_1.FILE,
    tags_1.FIRES,
    tags_1.TEMPLATE,
    tags_1.TYPE_PARAM,
    tags_1.FUNCTION,
    tags_1.NAMESPACE,
    tags_1.BORROWS,
    tags_1.CLASS,
    tags_1.EXTENDS,
    tags_1.MEMBER,
    tags_1.TYPEDEF,
    tags_1.TYPE,
    tags_1.PROPERTY,
    tags_1.CALLBACK,
    tags_1.PARAM,
    tags_1.YIELDS,
    tags_1.RETURNS,
    tags_1.THROWS,
    "other",
    tags_1.SEE,
    tags_1.TODO,
];
exports.TAGS_ORDER = TAGS_ORDER;
