// eslint-plugin-regexp/dist/utils/regexp-ast/is-equals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEqualNodes = void 0;
function isEqualChar(a, b, toCharSet) {
    if (a.type === "Character") {
        if (b.type === "Character") {
            if (a.value === b.value) {
                return true;
            }
        }
        else if (b.type === "CharacterSet") {
            return false;
        }
    }
    else if (a.type === "CharacterSet") {
        if (b.type === "Character") {
            return false;
        }
        else if (b.type === "CharacterSet") {
            return a.raw === b.raw;
        }
    }
    else if (a.type === "CharacterClassRange") {
        if (b.type === "CharacterClassRange") {
            return a.min.value === b.min.value && a.max.value === b.max.value;
        }
    }
    if (a.raw === b.raw) {
        return true;
    }
    return toCharSet(a).equals(toCharSet(b));
}
const EQUALS_CHECKER = {
    Alternative(a, b, toCharSet, shortCircuit) {
        return isEqualElements(a.elements, b.elements, toCharSet, shortCircuit);
    },
    Assertion(a, b, toCharSet, shortCircuit) {
        if (a.kind === "start" || a.kind === "end") {
            return a.kind === b.kind;
        }
        if (a.kind === "word") {
            return b.kind === "word" && a.negate === b.negate;
        }
        if (a.kind === "lookahead" || a.kind === "lookbehind") {
            if (b.kind === a.kind && a.negate === b.negate) {
                return isEqualAlternatives(a.alternatives, b.alternatives, toCharSet, shortCircuit);
            }
            return false;
        }
        return false;
    },
    Backreference(a, b) {
        return a.ref === b.ref;
    },
    CapturingGroup(a, b, toCharSet, shortCircuit) {
        return (a.name === b.name &&
            isEqualAlternatives(a.alternatives, b.alternatives, toCharSet, shortCircuit));
    },
    Character(a, b, toCharSet) {
        return isEqualChar(a, b, toCharSet);
    },
    CharacterClass(a, b, toCharSet) {
        return isEqualChar(a, b, toCharSet);
    },
    CharacterClassRange(a, b, toCharSet) {
        return isEqualChar(a, b, toCharSet);
    },
    CharacterSet(a, b, toCharSet) {
        return isEqualChar(a, b, toCharSet);
    },
    Flags(a, b) {
        return (a.dotAll === b.dotAll &&
            a.global === b.global &&
            a.ignoreCase === b.ignoreCase &&
            a.multiline === b.multiline &&
            a.sticky === b.sticky &&
            a.unicode === b.unicode);
    },
    Group(a, b, toCharSet, shortCircuit) {
        return isEqualAlternatives(a.alternatives, b.alternatives, toCharSet, shortCircuit);
    },
    Pattern(a, b, toCharSet, shortCircuit) {
        return isEqualAlternatives(a.alternatives, b.alternatives, toCharSet, shortCircuit);
    },
    Quantifier(a, b, toCharSet, shortCircuit) {
        return (a.min === b.min &&
            a.max === b.max &&
            a.greedy === b.greedy &&
            isEqualNodes(a.element, b.element, toCharSet, shortCircuit));
    },
    RegExpLiteral(a, b, toCharSet, shortCircuit) {
        return (isEqualNodes(a.pattern, b.pattern, toCharSet, shortCircuit) &&
            isEqualNodes(a.flags, b.flags, toCharSet, shortCircuit));
    },
};
function isToCharSetElement(node) {
    return (node.type === "Character" ||
        node.type === "CharacterClass" ||
        node.type === "CharacterClassRange" ||
        node.type === "CharacterSet");
}
function isEqualNodes(a, b, toCharSet, shortCircuit) {
    if (isToCharSetElement(a) && isToCharSetElement(b)) {
        return isEqualChar(a, b, toCharSet);
    }
    if (a.type !== b.type) {
        return false;
    }
    if (shortCircuit) {
        const kind = shortCircuit(a, b);
        if (kind != null) {
            return kind;
        }
    }
    if (/[(*+?[\\{|]/.test(a.raw) || /[(*+?[\\{|]/.test(b.raw)) {
        return EQUALS_CHECKER[a.type](a, b, toCharSet, shortCircuit);
    }
    return a.raw === b.raw;
}
exports.isEqualNodes = isEqualNodes;
function isEqualElements(a, b, toCharSet, shortCircuit) {
    if (a.length !== b.length) {
        return false;
    }
    for (let index = 0; index < a.length; index++) {
        const ae = a[index];
        const be = b[index];
        if (!isEqualNodes(ae, be, toCharSet, shortCircuit)) {
            return false;
        }
    }
    return true;
}
function isEqualAlternatives(a, b, toCharSet, shortCircuit) {
    if (a.length !== b.length) {
        return false;
    }
    const beList = [...b];
    for (const ae of a) {
        const bIndex = beList.findIndex((be) => isEqualNodes(ae, be, toCharSet, shortCircuit));
        if (bIndex >= 0) {
            beList.splice(bIndex, 1);
        }
        else {
            return false;
        }
    }
    return true;
}
